Adding authorisation options for API access

bpiraeus opened the pull request:
Fixes #

## Requirements

All new code should be covered with tests, documentation should be updated. CI should pass.

## Description of the Change

<!--

Adds authorisation options into aptly for the API and local repositories, currently only ldap auth is available however it is designed to accept other methods without a lot of additional work.

allows for per-repository ldap group authorisation ensuring only those in the correct group may muddle with the repo.

-->

## Checklist

- [ ] unit-test added (if change is algorithm)
- [ ] functional test added/updated (if change is functional)
- [ ] man page updated (if applicable)
- [ ] bash completion updated (if applicable)
- [ ] documentation updated
- [ ] author name in `AUTHORS`

bpiraeus committed:
Adding authorisation options for API access
  - ldap currently the only supported method
adding authorisation options for local repositories
  - ldap groups per repo

bpiraeus commented:
I went through the failing tests, and unless I'm missing something, those are not from my changes, not sure what I can do about that short of refactoring someone else's code. I'm happy to muck around if needed, but you'd likely prefer someone with a lot more chops than myself to mitigate those issues.

randombenj commented:
@bpiraeus It looks like mostly deprecation errors 

- https://github.com/golang/go/issues/44226
- package golang.org/x/crypto/ssh/terminal is deprecated: this package moved to golang.org/x/term

If you want to, feel free to investigate if there are some simple fixes for the deprecated packages (x/term) seems to be quite an easy one, not sure about the crypto one ...

bpiraeus committed:
minor cleanup based on make check

bpiraeus commented:
Cleaned up a few oddballs in my code that gosimple flagged, also moved the terminal bits to the appropriate package, the pgp ones appear as if they're likely going to want 3rd party library.

randombenj commented:
@bpiraeus I just re reviewd the changes in `go.mod` and it seems like you upgraded quite a lot of dependencies. Was this on purpose? I think if you just add the packages you need to add without upgrading the others it will 'fix' the linting.

I would propose to upgrade all the other deps in a sepparate change. What do you think?

neolynx commented:
introducing authorization per repo is a nice idea. 

however, authentication without SSL is not that nice, so probably we should implement this as well with this change. 

since different authentication types are supported, the variable `LdapGroup` should probably `AuthorizedGroup`

neolynx commented:
rebased on master

neolynx commented:
@bpiraeus are you still interested in continuing this ?
